File: PythonAlgorithmInterView/그래프/조합의 합.py
def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:

    result = []

    def dfs(csum, index, path):
        if csum < 0 :
            return
        if csum == 0 :
            result.append(path)
            return

        for i in range(index, len(candidates)):
            dfs(csum - candidates[i], i, path + [candidates[i]] )

    dfs(target, 0, [])

    return result

if __name__ == '__main__':
    candidates = [2, 3, 6, 7]
    target = 7

    combinationSum(None, candidates, target)

File: PythonAlgorithmInterView/그래프/조합의 합.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 * 문제 39번
 */
 var combinationSum = function(candidates, target) {
    
    let result = [];

    function dfs(csum, index, path){

        if( csum < 0 ){
            return;
        }
        if ( csum == 0 ){
            result.push(path)
        }

        for( let i = index; i < candidates.length; i++){
            dfs( csum - candidates[i], i, path.concat(candidates[i]));
        }
    }

    dfs(target, 0, [])

    return result;
};

let candidates = [2,3,6,7], target = 7;

console.log(combinationSum(candidates, target))
